AA-543: Refactor post execution

drortirosh opened the pull request.

drortirosh committed:
refactor "callPostOp" from "postExecution"

drortirosh committed:
cleanups

drortirosh committed:
Merge branch 'develop' into refactor-post-execution

drortirosh committed:
lints

drortirosh committed:
misc

drortirosh committed:
gascalcs

drortirosh committed:
refactor postInnerCall

drortirosh committed:
cleanups

drortirosh committed:
inner return success, gasused

drortirosh committed:
bugfix gas on error

drortirosh committed:
fix gas on error (penalty)

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -113,47 +116,68 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
-            }
+        bool callSuccess;
+        if (success) {
+            callSuccess = returnData != 0;
```
I would suggest `executionSuccess` or something like that, more explicit.

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -113,47 +116,68 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
-                emitUserOperationEvent(opInfo, false, actualGasCost, actualGas);
-                collected = actualGasCost;
+                // innerCall reverted on prefund too low.
+                // below we re-calculate the cast cost and refund.
```
Where? Put function name, lines or some other meaningful reference to what you mean here.

Also, this entire `if` is empty, could we maybe "invert" it to be `else if (!innerRevertCode == INNER_REVERT_LOW_PREFUND) { ... } ` ??

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -113,47 +116,68 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
+    }
+
+    /**
+     * call after innerCall executed to call the account's callData and paymaster's postOp.
+     * - calculate the overall gas used and unuased penalty.
+     * - refund payer if needed
+     * - emit UserOperationEvent
```
Rewrite this comment. Use ChatGPT.

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -113,47 +116,68 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
+        unchecked {
+            uint256 prefund = opInfo.prefund;
+            uint256 actualGasCost = actualGas * getUserOpGasPrice(opInfo.mUserOp);
+            uint256 refund;
```
Are you seriously creating two variables - `refund` and `prefund` here? Absolutely not. These two variables MUST have two or three words names that don't sound anything alike.

drortirosh replied:
suggest better names.

forshtat replied:
I don't know, `originalPrefund` and `remainingGasRefund`?
@shahafn could you help me out here? 😄

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -113,47 +116,68 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
+                refund = prefund - actualGasCost;
+            } else {
+                actualGasCost = prefund;
+                //depending where the over-gas-used was found, we either reverted innerCall or not.
```
Instead of writing this comment add a boolean parameter to the `emitPrefundTooLow` - we don't always need it but it will help make the code flow more clear.

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -356,9 +381,11 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
                 INNER_GAS_OVERHEAD
             ) {
+                uint256 gasUsed = preGas - gasleft();
                 assembly ("memory-safe") {
                     mstore(0, INNER_OUT_OF_GAS)
-                    revert(0, 32)
+                    mstore(32, gasUsed)
+                    revert(0, 64)
```
If we are making a  big refactoring now, would you consider extracting this entire `unchecked { ... }` block to some `_verifyEnoughGasForInner` function? This is a very distracting code.

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -753,75 +781,60 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
-                    uint256 postOpGasUsed = postOpPreGas - gasleft();
-                    postOpUnusedGasPenalty = _getUnusedGasPenalty(postOpGasUsed, mUserOp.paymasterPostOpGasLimit);
+            actualGas = executionGas + _getUnusedGasPenalty(executionGas, mUserOp.callGasLimit) + opInfo.preOpGas;
+            if (context.length > 0) {
```
Also for the entire `if (context.length > 0) { ... }` block - it would probably cost nothing to make it an internal function.

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -753,75 +781,60 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
+            if (context.length > 0) {
+                uint256 postOpPreGas = gasleft();
+                uint256 actualGasCostForPostOp = actualGas * gasPrice;
+                uint256 paymasterPostOpGasLimit = mUserOp.paymasterPostOpGasLimit;
```
Why create a variable if `mUserOp` is in memory anyways? MLOAD costs 3 gas, no?

drortirosh replied:
removed.
mUserOp is "double-reference", and in the past, it was better to copy a struct member to a local (stack) one.
but apparently, the compiler now behaves better if we don't.
(btw: its not good optimization of the compiler: it should be agnostic to the way we read again the same memory var)

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -753,75 +781,60 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
+                uint256 postOpPreGas = gasleft();
+                uint256 actualGasCostForPostOp = actualGas * gasPrice;
+                uint256 paymasterPostOpGasLimit = mUserOp.paymasterPostOpGasLimit;
+                // todo: no real need to check if paymaster has code. we already called validatePaymasdterUserOp
```
Now is not a time to add new TODOs. Either go full assembly (not recommened) or ignore the unnecessary "has code" checks you get from solidity (recommended).

forshtat commented on contracts/core/EntryPoint.sol:
```diff
@@ -753,75 +781,60 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
+                // todo: no real need to check if paymaster has code. we already called validatePaymasdterUserOp
+                try IPaymaster(paymaster).postOp{
+                        gas: paymasterPostOpGasLimit
+                    }(mode, context, actualGasCostForPostOp, gasPrice)
```
We can totally remove the `postOpReverted` "mode" from the codebase by the way.

drortirosh committed:
pr comments

drortirosh committed:
fix test

drortirosh committed:
remove obsolete opReverted enum value

drortirosh committed:
pr chnages

drortirosh committed:
extract _callPostOp

note: most of the gas reduction are due to change of context to
"calldata" from "memory", not from the method extract...

drortirosh committed:
lint

drortirosh committed:
remove obsolete comment

drortirosh committed:
minor fixes

drortirosh committed:
remove "else", add comment

forshtat committed:
Merge branch 'develop' into refactor-post-execution

drortirosh committed:
gascalc

forshtat commented:
A remider to self to remove `postOpReverted` from ERC-4337 if this PR is merged.

drortirosh commented:
Deferred
